lib/wallet/src/Cardano/Wallet/Transaction.hs

```diff
@@ -357,6 +364,71 @@ data DelegationAction
 instance Buildable DelegationAction where
     build = genericF
 
+data DepositEvent = DepositTaken !Coin | DepositReturned !Coin
+    deriving (Eq, Generic, Show)
+
+data DepositEventParams = DepositEventParams
+    { protocolParameterDeposit :: !Coin
+    , txResolvedInputs :: ![Coin]
+    , txResolvedCollateralInputs :: ![Coin]
+    , txOutputs :: ![Coin]
+    , txCollateralOutputs :: ![Coin]
+    , txFee :: !Coin
+    , txWithdrawals :: ![Coin]
+    } deriving (Eq, Generic, Show)
+
+-- Given information about a particular transaction (represented by
+-- the 'DepositEventParams'), determine whether a deposit was taken or
+-- a refund was issued.
+--
+-- Caveat #1: it is possible (though unlikely) that transaction registers or
+-- unregisters a stake key associated with a wallet A, at the same time
+-- containing payments to the wallet B, thus also being associated with B.
+-- In this case, the deposit event might be incorrectly attributed to B.
+--
+-- Caveat #2: Ideally we only want to detect deposits/refunds if
+-- a key registration/un-registration certificate is present in the transaction.
+-- Its not possible to rely on this information yet, as it wasn't historically
+-- recorded in the database.
+--
+-- Caveat #3: The implementation is somewhat fragile, as it relies on the
+-- assumption that outside of the inputs/outputs, collateral inputs/outputs,
+-- fees and withdrawals there are no explanations other than deposits/redunds
+-- for a posible discrepancy between total input and total output amounts.
+-- Moreover, if such discrepancy is not equal to the deposit amount, then
+-- we assume that no deposit/refund event took place.
+--
+-- Caveat #4: There is a corner case: if a transaction contains
+-- a key de-registration certificate + refund only
+-- then its considered "Incoming" as it's increases the wallet balance.
+-- However, the wallet doesn't record fees for "incoming" transactions ¯\_(ツ)_/¯
+-- And in this case the fee is 0.
+determineDepositEvent :: DepositEventParams -> Maybe DepositEvent
```

Comment:
what about adding property tests or unit testing  this function? In the later case you could at least capture/document corner cases through unit tests...